misc/python/materialize/checks/all_checks/alter_connection.py

```diff
@@ -0,0 +1,276 @@
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+from textwrap import dedent
+
+from materialize.checks.actions import Testdrive
+from materialize.checks.checks import Check
+from materialize.checks.common import KAFKA_SCHEMA_WITH_SINGLE_STRING_FIELD
+
+
+def schema() -> str:
+    return dedent(KAFKA_SCHEMA_WITH_SINGLE_STRING_FIELD)
+
+
+class AlterConnectionToSsh(Check):
+    def initialize(self) -> Testdrive:
+        return Testdrive(
+            schema()
+            + dedent(
+                """
+                $ kafka-create-topic topic=alter-connection-1
+
+                $ kafka-ingest topic=alter-connection-1 format=bytes
+                one
+
+                > CREATE CONNECTION kafka_conn_alter_connection_1
+                  TO KAFKA (BROKER '${testdrive.kafka-addr}');
+
+                > CREATE SOURCE alter_connection_source_1
+                  FROM KAFKA CONNECTION kafka_conn_alter_connection_1 (TOPIC 'testdrive-alter-connection-1-${testdrive.seed}')
+                  FORMAT TEXT
+                  ENVELOPE NONE;
+                """
+            )
+        )
+
+    def manipulate(self) -> list[Testdrive]:
+        return [
+            Testdrive(schema() + dedent(s))
+            for s in [
+                """
+                $ kafka-ingest topic=alter-connection-1 format=bytes
+                two
+
+                > ALTER CONNECTION kafka_conn_alter_connection_1 SET BROKER '${testdrive.kafka-addr}' USING SSH TUNNEL thancred;
+
+                $ kafka-ingest topic=alter-connection-1 format=bytes
+                three
+
+                $ kafka-create-topic topic=alter-connection-2
+
+                $ kafka-ingest topic=alter-connection-2 format=bytes
+                ten
+
+                > CREATE CONNECTION kafka_conn_alter_connection_2
+                  TO KAFKA (BROKER '${testdrive.kafka-addr}');
+
+                > CREATE SOURCE alter_connection_source_2
+                  FROM KAFKA CONNECTION kafka_conn_alter_connection_2 (TOPIC 'testdrive-alter-connection-2-${testdrive.seed}')
+                  FORMAT TEXT
+                  ENVELOPE NONE;
+
+                $ kafka-ingest topic=alter-connection-2 format=bytes
+                twenty
+                """,
+                """
+                $ kafka-ingest topic=alter-connection-1 format=bytes
+                four
+
+                $ kafka-ingest topic=alter-connection-2 format=bytes
+                thirty
+
+                > ALTER CONNECTION kafka_conn_alter_connection_2 SET BROKER '${testdrive.kafka-addr}' USING SSH TUNNEL thancred;
+
+                $ kafka-ingest topic=alter-connection-2 format=bytes
+                fourty
+                """,
+            ]
+        ]
+
+    def validate(self) -> Testdrive:
+        return Testdrive(
+            dedent(
+                """
+                > SELECT * FROM alter_connection_source_1;
```

Comment:
Ideally in validate there would also be a way of checking that the connection is actually altered as expected. Right now if the alters are not persisted the check still succeeds, right?
I see two ways:
1. Destroy the old connection
2. Check the metadata of the connection

Reply:
I haven't achieved to check that yet.

> 1. Destroy the old connection

I believe there is no old connection (accessible to me) since I change an existing connection.

> 2. Check the metadata of the connection

Not sure if I miss something but at least `mz_connections`, `mz_kafka_connections`, and `mz_ssh_tunnel_connections` do not expose anything useful for me.

Any suggestions?

Reply:
The only way to check that it has changed in the expected way is to:
- Know what the correct values are for the connection--break them, assert that you cannot connect, then fix them and validate that you can connect.
- Look in the catalog via `SHOW CREATE CONNECTION` and demo that the values are durably recorded as different.

We could also test that the alters are durably recorded by creating the connection with invalid credentials and `WITH (VALIDATE = false)`, fixing them, then restarting MZ and ensuring things work.

Reply:
Thank you, I [have added that!](https://github.com/MaterializeInc/materialize/pull/23076/commits/b7c7c5e6b85c232f01416ce50a99ed867d2e27bc)